File: src/main/kotlin/chapter06/ScoreCollection.java
package chapter06;

import java.util.*;

public class ScoreCollection {
   private List<Scoreable> scores = new ArrayList<>();
   
   public void add(Scoreable scoreable) {
      if (scoreable == null) throw new IllegalArgumentException();   // 입력 범위를 분명하게 하기
      scores.add(scoreable);
   }
   
   public int arithmeticMean() {
      if (scores.size() == 0) return 0; // ArithmeticException 발생 가능성
      long total = scores.stream().mapToLong(Scoreable::getScore).sum();
      return (int) (total / scores.size());
   }
}

File: src/main/kotlin/chapter06/ScoreCollection.kt
package chapter06


class ScoreCollection {
    private val scores: MutableList<Scoreable> = ArrayList()
    fun add(scoreable: Scoreable) {
        scores.add(scoreable)
    }

    fun arithmeticMean(): Int {
        if (scores.size == 0) return 0 // ArithmeticException 발생 가능성
        val total = scores.stream().mapToLong { obj: Scoreable -> obj.getScore().toLong() }.sum()
        return (total / scores.size).toInt()
    }
}

fun interface Scoreable {
    fun getScore(): Int
}